File: prevention_field/prevention_field/report/customer_item_matrix/customer_item_matrix.js
// Copyright (c) 2024, ahmed waleed and contributors
// For license information, please see license.txt

frappe.query_reports["Customer Item Matrix Report"] = {
    "filters": [
        {
            "label": __("Salesperson"),
            "fieldname": "salesperson",
            "fieldtype": "Link",
            "options": "Sales Person",
            "default": "",
            "reqd": 1
        },
        {
            "label": __("From Date"),
            "fieldname": "from_date",
            "fieldtype": "Date",
            "default": "",
            "reqd": 1
        },
        {
            "label": __("To Date"),
            "fieldname": "to_date",
            "fieldtype": "Date",
            "default": "",
            "reqd": 1
        }
    ],

    "get_data": function (filters) {
        return frappe.call({
            method: "frappe.client.get_list",
            args: {
                doctype: "Sales Invoice",
                fields: ["name"],
                filters: {
                    docstatus: 1,
                    custom_collect_status: "UnCollected", // Added filter for custom_collect_status
                    ...(filters.from_date ? {posting_date: [">=", filters.from_date]} : {}),
                    ...(filters.to_date ? {posting_date: ["<=", filters.to_date]} : {}),
                    ...(filters.salesperson ? {sales_person: filters.salesperson} : {})
                },
                limit_page_length: 1000
            }
        }).then(response => {
            let invoices = response.message || [];
            let invoice_names = invoices.map(invoice => invoice.name);

            if (invoice_names.length === 0) {
                return [];
            }

            return frappe.call({
                method: "frappe.client.get_list",
                args: {
                    doctype: "Sales Invoice Item",
                    fields: ["parent", "item_code", "qty"],
                    filters: {
                        parent: ["in", invoice_names],
                        docstatus: 1
                    },
                    limit_page_length: 1000
                }
            }).then(response => {
                let sales_data = response.message || [];
                let data = [];
                let customer_map = {};

                sales_data.forEach(item => {
                    let customer = get_customer(item.parent);
                    if (!customer_map[customer]) {
                        customer_map[customer] = {};
                    }
                    if (!customer_map[customer][item.item_code]) {
                        customer_map[customer][item.item_code] = 0;
                    }
                    customer_map[customer][item.item_code] += item.qty;
                });

                for (let [customer, items] of Object.entries(customer_map)) {
                    let row = {customer: customer};
                    for (let [item, qty] of Object.entries(items)) {
                        row[item] = qty;
                    }
                    data.push(row)
                }

                return data;
            }).catch(error => {
                console.error("Error fetching sales invoice items:", error);
                return [];
            });
        }).catch(error => {
            console.error("Error fetching sales invoices:", error);
            return [];
        });
    }
};

async function get_customer(invoice_name) {
    try {
        let response = await frappe.call({
            method: "frappe.client.get_list",
            args: {
                doctype: "Sales Invoice",
                fields: ["customer"],
                filters: {
                    name: invoice_name
                },
                limit_page_length: 1
            }
        });

        let customer_list = response.message || [];
        return customer_list.length > 0 ? customer_list[0].customer : "Unknown";
    } catch (error) {
        console.error("Error fetching customer:", error);
        return "Unknown";
    }
}

File: prevention_field/prevention_field/report/customer_item_matrix/customer_item_matrix.py
# Copyright (c) 2024, ahmed waleed and contributors
# For license information, please see license.txt

import frappe
from frappe import _

CUSTOM_COLLECT_STATUS = 'UnCollected'
REQUIRED_FILTERS = ["salesperson", "from_date", "to_date"]


def execute(filters=None):
    if not filters:
        filters = {}

    # Validate the required filters are present
    validate_filters(filters)

    columns = get_columns(filters)
    data = get_data(filters)

    return columns, data


def validate_filters(filters):
    """Validate that all required filters are provided."""
    for f in REQUIRED_FILTERS:
        if not filters.get(f):
            frappe.throw(_("The {0} filter is mandatory").format(frappe.bold(f.title())))


def get_columns(filters):
    """Dynamically define columns based on items in sales invoices."""
    # Get distinct item codes from sales invoices based on provided filters
    item_columns = frappe.db.sql("""
        SELECT DISTINCT sii.item_code
        FROM `tabSales Invoice Item` sii
        JOIN `tabSales Invoice` si ON sii.parent = si.name
        WHERE si.docstatus = 1 AND si.outstanding_amount = 0 
        AND si.is_return = 0 AND si.custom_collect_status = %(custom_collect_status)s
    """, {"custom_collect_status": CUSTOM_COLLECT_STATUS}, as_dict=True)

    columns = [
        {"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 150}
    ]

    for item in item_columns:
        columns.append({"label": item['item_code'], "fieldname": item['item_code'], "fieldtype": "Float", "width": 150})

    return columns


def get_data(filters):
    """Fetch sales data based on the filters and transform it into a matrix format."""
    conditions = []
    params = {}

    if filters.get("from_date"):
        conditions.append("si.posting_date >= %(from_date)s")
        params["from_date"] = filters["from_date"]
    if filters.get("to_date"):
        conditions.append("si.posting_date <= %(to_date)s")
        params["to_date"] = filters["to_date"]
    if filters.get("salesperson"):
        conditions.append("st.sales_person = %(salesperson)s")
        params["salesperson"] = filters["salesperson"]

    conditions.append("si.outstanding_amount = 0")
    conditions.append("si.custom_collect_status = %(custom_collect_status)s")
    params["custom_collect_status"] = CUSTOM_COLLECT_STATUS

    condition_string = " AND ".join(conditions) if conditions else "1=1"

    query = f"""
        SELECT
            si.customer_name AS customer,
            sii.item_code AS item,
            SUM(sii.qty) AS quantity_sold
        FROM
            `tabSales Invoice Item` sii
        JOIN
            `tabSales Invoice` si ON sii.parent = si.name
        JOIN
            `tabSales Team` st ON si.name = st.parent
        WHERE
            si.docstatus = 1 AND  si.is_return = 0  AND {condition_string}
        GROUP BY
            si.customer, sii.item_code
    """

    try:
        results = frappe.db.sql(query, params, as_dict=True)
    except Exception as e:
        frappe.log_error(_("Error fetching data: {0}").format(str(e)), "Customer item matrix report")
        return []

    return transform_to_matrix(results)


def transform_to_matrix(results):
    """Transform the results into a customer-item matrix format."""
    from collections import defaultdict
    matrix = defaultdict(lambda: defaultdict(float))

    # Aggregate data by customer and item
    for row in results:
        matrix[row['customer']][row['item']] += row['quantity_sold']

    data = []
    # Prepare customer-wise data
    for customer, items in matrix.items():
        row = {"customer": customer}
        for item, qty in items.items():
            row[item] = qty
        data.append(row)

    return data
